### PR TITLE
#19231 Fix exception at pasteelementat

### LINKED ISSUE #19231: [d3412c] openrct2.exe: OpenRCT2::TileInspector::PasteElementAt
<table>
    <tr>
        <td></td>
        <td><b><a href="https://openrct2.sp.backtrace.io/dashboard/openrct2/project/openrct2/group/d3412cb50b962480c7c4f60eb29251e383bee1f085bf920ee6f573e4442caa8f">New error: [openrct2]</a></b><br /><b>1</b> occurrence since (Fri Nov 15 2019 11:27:39)<br /></td>
    </tr>
</table><table>
    <tr><td><strong>Application</strong></td><td>openrct2.exe</td></tr>
<tr>
    <td><strong>Attributes</strong></td>
    <td>
<strong>_rxid</strong> &mdash; 1
        <ul>
                <li><code>38000000-484c-7e09-0000-000000000000</code> &mdash; 1</li></ul><br>
<strong>attachment_config.ini</strong> &mdash; 1
        <ul>
                <li><code>C:\Users\keest\Documents\OpenRCT2\crash\f0811dbe-0044-466e-b9a4-e61c085055d2.ini</code> &mdash; 1</li></ul><br>
<strong>assert_failure</strong> &mdash; 1
        <ul>
                <li><code>*</code> &mdash; 1</li></ul><br>
<strong>fault.address</strong> &mdash; 1
        <ul>
                <li><code>0</code> &mdash; 1</li></ul><br>
<strong>version</strong> &mdash; 1
        <ul>
                <li><code>OpenRCT2, v0.4.3-160-g3456797 (3456797 on develop) provided by GitHub</code> &mdash; 1</li></ul><br>
<strong>error.message</strong> &mdash; 1
        <ul>
                <li><code>EXCEPTION_ACCESS_VIOLATION_READ</code> &mdash; 1</li></ul><br>
<strong>commit</strong> &mdash; 1
        <ul>
                <li><code>3456797</code> &mdash; 1</li></ul><br></td>
    </tr>
    <tr>
        <td><strong>Classifiers</strong></td>
        <td>
            <strong>invalid-read</strong><br></td>
    </tr>
    <tr>
        <td><strong>Frames</strong></td>
        <td>
            <code>
OpenRCT2::TileInspector::PasteElementAt</code>
        </td>
    </tr></table>

### PR DESCRIPTION
**Pull request by jayjay300**
Attempt to fix #19231 due to read error

### CODE REVIEW

File: src/openrct2/world/TileInspector.cpp [hunk 1]
```diff
@@ -343,7 +343,7 @@ namespace OpenRCT2::TileInspector
                     return GameActions::Result(GameActions::Status::Unknown, STR_TOO_MANY_BANNERS_IN_GAME, STR_NONE);
                 }
                 // Copy the banners style
-                *newBanner = *GetBanner(bannerIndex);
+                newBanner = GetBanner(bannerIndex);
```
Comment (ZehMatt) [inline src/openrct2/world/TileInspector.cpp:346]: This is no longer correct, the purpose is to actually copy the data and not change pointer

**Review by ZehMatt — CHANGES_REQUESTED** [review]

File: src/openrct2/world/TileInspector.cpp (hunk 1 above)
Comment (duncanspumpkin) [inline src/openrct2/world/TileInspector.cpp:346]: This isn't what you want to do. You would need to verify that the bannerIndex is creating a valid banner. I assume that GetBanner(bannerIndex) is returning nullptr. You just need to check for that and if it happens then return a failure. If you want to be extra helpful you would do this before the `if (isExecuting)`

Comment (jayjay300) [reply src/openrct2/world/TileInspector.cpp:346]: I understand now. Thanks for the feedback, I will adjust accordingly. Yeah GetBanner is returning a nullptr so will check for that.

**Comment by jayjay300** [comment]
@Broxzier I think I understand what you're saying here, but a bit unsure of how to go about that implementation and where to place this. Should it be a new object type that holds both the banner entry and copied element?

**Comment by jayjay300** [comment]
@Broxzier Appreciate the explanation, it was indeed super helpful. I think the best option right now would be to merge the current solution and maybe open another issue for this rework -- I agree I also am not sure how the copied element is retained once the window is closed and this looks to be a bigger issue than previously thought. 
I took this issue as it was marked as good first issue, still finding my way around the code and might need some more experience learning the codebase before I can tackle this fully. I will circle back and reattempt in the future or someone else can pick it up in the meantime.

**Comment by Broxzier** [comment]
Alright jayjay, thanks for replying. In that case, this PR looks good and is ready to be merged.

**Comment by jayjay300** [comment]
@Broxzier rebased and added fix to changelog -- my name is already in the contributors list for bug fixes from the last release actually so no need to add. Let me know if there's anything else that needs to be updated, thanks for reviews on this!

**Comment by ZehMatt** [comment]
Your commit history looks a bit off, can you try fixing it by rebasing onto the latest develop branch?  Ordinarily this should be enough to do:
```
git fetch upstream develop
git rebase upstream/develop
```
And then you need to force push

**Comment by jayjay300** [comment]
@ZehMatt followed that and looks like it worked, let me know if there's any more issues

**Comment by ZehMatt** [comment]
It looks better now but for some reason you still have unrelated commits included. Perhaps its best to just reset to origin/develop and apply the changes from scratch, given the diff it should be fairly simple and keeps the history clean. So the steps would be following:
```
git fetch upstream develop
git reset --hard upstream/develop
```
apply the changes and commit, after that you have to force push again and then it should be a single commit.

File: distribution/changelog.txt [hunk 2]
```diff
@@ -3,6 +3,7 @@
 - Fix: [#19296] Crash due to a race condition for parallel object loading.
 - Fix: [#19756] Crash with title sequences containing no commands.
 - Fix: [#19767] No message when path is not connected to ride exit and is therefore unreachable for mechanics.
+- Fix: [#19231] Crash due to null pointer to previously deleted banner in tile copy/paste functionality
```
Comment (Broxzier) [inline distribution/changelog.txt:6]: This should be moved up all the way to the top, so that the issue/pr numbers are incremental.

File: src/openrct2/world/TileInspector.cpp [hunk 3]
```diff
@@ -1065,4 +1070,4 @@ namespace OpenRCT2::TileInspector
     {
         return _highlightedElement == elem;
     }
-} // namespace OpenRCT2::TileInspector
+} // namespace OpenRCT2::TileInspector
```
Comment (Broxzier) [inline src/openrct2/world/TileInspector.cpp:1073]: Unrelated change, please revert:
```suggestion
} // namespace OpenRCT2::TileInspector

```

**Comment by Broxzier** [comment]
Unfortuantely, there is another merge conflict now, since another changelog entry was added in the meantime. Could you rebase again?

**Comment by jayjay300** [comment]
Sorry didn't mean to close this-- must've closed automatically with the rebase. Let me know if any further adjustments are needed
